chore(#9621): add a @medic/constants shared library

cliftonmcintosh opened the pull request:
<!--
Please use semantic PR titles that respect this format:

<type>(#<issue number>): <subject>

Quick example:

feat(#1234): add hat wobble
^--^(#^--^): ^------------^
|     |      |
|     |      + - > subject
|     |
|     + -------- > issue number
|
+ -------------- > type: chore, feat, fix, perf.

https://docs.communityhealthtoolkit.org/contribute/code/workflow/#commit-message-format
-->

# Description

<!-- DESCRIPTION -->

This adds and uses a @medic/constants shared library.

<!-- ISSUE NUMBER -->

Closes #9621 

## Details

This converts several magic strings into constants. It does not yet include all of the magic strings that were mentioned in #9621, but I wanted to get feedback before proceeding further.

# Topics and questions for draft review

## Is "chore" the right type for this change?

It seems to me that this change doesn't fit any of the other categories defined in the [commit message format](https://docs.communityhealthtoolkit.org/community/contributing/code/workflow/#commit-message-format) documentation. It feels like "chore" is the closest category, but "chore" was defined as "Non-code" in that documentation. Please let me know if I should assign this change to a different category.

## Should this be broken into parts?

Would smaller commits with focused changes on each of the constant be preferred? For example, one PR for `'mm-online'`, another for `'service-worker-meta'`, and so on.

## `npm run build-dev` throws an error when the constants are used in `admin/src/js/services/session.js`

When I run `npm run build-dev` on this branch, I get an error with StringMap: `Error: StringMap expected string key`. I don't get this error on master.

This is also failing the compile step in github: 
https://github.com/medic/cht-core/actions/runs/19305186560/job/55212091760?pr=10457

Any guidance on what change might cause that error would be appreciated.

<details>

<summary>Here is a snippet of the terminal output with the error</summary>

```sh
build-prepare: building admin app
Build angularjs template cache from "admin/src/templates/" to "api/build/static/admin/js/templates.js"
Appending template: authorization.html
Appending template: authorization_permissions.html
Appending template: authorization_roles.html
Appending template: backup.html
Appending template: delete_doc_confirm.html
Appending template: delete_user_confirm.html
Appending template: display.html
Appending template: display_date_time.html
Appending template: display_languages.html
Appending template: display_privacy_policies.html
Appending template: display_privacy_policies_preview.html
Appending template: display_translations.html
Appending template: edit_language.html
Appending template: edit_translation.html
Appending template: edit_user.html
Appending template: export.html
Appending template: export_contacts.html
Appending template: export_dhis.html
Appending template: export_feedback.html
Appending template: export_messages.html
Appending template: export_reports.html
Appending template: export_user-devices.html
Appending template: forms_xml.html
Appending template: header.html
Appending template: images.html
Appending template: images_branding.html
Appending template: images_icons.html
Appending template: images_partners.html
Appending template: images_tabs_icons.html
Appending template: import_translation.html
Appending template: message_queue.html
Appending template: message_queue_tab.html
Appending template: modal.html
Appending template: multiple_user_modal.html
Appending template: pagination.html
Appending template: release.html
Appending template: sms.html
Appending template: sms_forms.html
Appending template: sms_settings.html
Appending template: sms_test.html
Appending template: targets.html
Appending template: targets_edit.html
Appending template: translation.html
Appending template: upgrade.html
Appending template: upgrade_abort.html
Appending template: upgrade_confirm.html
Appending template: users.html
browserify-admin: making dir
browserify-admin: running browserify on main.js
/Users/cliftonmcintosh/Documents/workspace/cht-core/node_modules/stringmap/stringmap.js:99
            throw new Error("StringMap expected string key");
            ^

Error: StringMap expected string key
    at stringmap.set (/Users/cliftonmcintosh/Documents/workspace/cht-core/node_modules/stringmap/stringmap.js:99:19)
    at Scope.add (/Users/cliftonmcintosh/Documents/workspace/cht-core/node_modules/ng-annotate/build/es5/scope.js:102:17)
    at /Users/cliftonmcintosh/Documents/workspace/cht-core/node_modules/ng-annotate/build/es5/scopetools.js:38:25
    at Array.forEach (<anonymous>)
    at createScopes (/Users/cliftonmcintosh/Documents/workspace/cht-core/node_modules/ng-annotate/build/es5/scopetools.js:36:27)
    at visit (/Users/cliftonmcintosh/Documents/workspace/cht-core/node_modules/ordered-ast-traverse/ordered-ast-traverse.js:25:19)
    at visit (/Users/cliftonmcintosh/Documents/workspace/cht-core/node_modules/ordered-ast-traverse/ordered-ast-traverse.js:43:25)
    at traverse (/Users/cliftonmcintosh/Documents/workspace/cht-core/node_modules/ordered-ast-traverse/ordered-ast-traverse.js:56:5)
    at Object.setupScopeAndReferences (/Users/cliftonmcintosh/Documents/workspace/cht-core/node_modules/ng-annotate/build/es5/scopetools.js:18:5)
    at ngAnnotate (/Users/cliftonmcintosh/Documents/workspace/cht-core/node_modules/ng-annotate/build/es5/ng-annotate-main.js:1123:16)
```

</details>

### UPDATE

I ran `git bisect` to try to pin this down, and the problem seems to be in commit 3e154fe55d718403250c598549eaf114cfa6b9a6. I'll experiment with changes to see exactly what might be causing the problem there.

### ANOTHER UPDATE

The problem is with using the constants library in `admin/src/js/services/session.js`. When I remove its use there, the build problem is resolved. Any idea why?

# Code review checklist
<!-- Remove or comment out any items that do not apply to this PR; in the remaining boxes, replace the [ ] with [x]. -->
- [ ] UI/UX backwards compatible: Test it works for the new design (enabled by default). And test it works in the old design, enable `can_view_old_navigation` permission to see the old design. Test it has appropriate design for RTL languages. 
- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

<!-- COMPOSE URLS GO HERE - DO NOT CHANGE -->

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

cliftonmcintosh commented on shared-libs/constants/package.json:
```diff
@@ -0,0 +1,14 @@
+  },
+  "scripts": {
+    "test": "nyc --nycrcPath='../nyc.config.js' mocha ./test"
+  },
```
Is this "scripts" section needed or necessary? `shared-libs/constants/index.js` is just a grouping of constants and there are no direct tests for them.

cliftonmcintosh replied:
Thanks. I've removed the "directories" and "scripts" keys from the JSON in this file

cliftonmcintosh commented:
It is [failing to compile](https://github.com/medic/cht-core/actions/runs/19375463728/job/55447106757?pr=10457). Any idea why?

It looks like https://github.com/medic/cht-core/pull/10462 is [failing at the same step for the same reason](https://github.com/medic/cht-core/actions/runs/19367721622/job/55415163633?pr=10462).

cliftonmcintosh commented on api/tests/mocha/controllers/changes.spec.js:
```diff
@@ -25,7 +26,7 @@ describe('Changes controller', () => {
+        { id: DOC_IDS.SERVICE_WORKER_META },
         { id: '_design/medic-client', },
         { id: 'org.couchdb.user:user', },
         { id: 'settings' },
```
Are these three strings candidates for constants?

`'org.couchdb.user:user'` only appears in tests. `'_design/medic-client'` appears in other files, not just tests.

cliftonmcintosh commented on tests/scalability/generate-real-world-data/generate-real-world-data.js:
```diff
@@ -120,7 +121,7 @@ const generatePerson = async (type, parents, isPrimaryContact) => {
   let roles = pairPlaceTypesRoles[type];
   if (isPrimaryContact && type === 'district_hospital') {
-    roles = ['national_admin', 'mm-online'];
+    roles = ['national_admin', USER_ROLES.ONLINE];
```
Is `'national_admin'` a candidate for a constant?

cliftonmcintosh commented on webapp/src/ts/services/db-sync.service.ts:
```diff
@@ -14,9 +14,10 @@ import { TranslateService } from '@mm-services/translate.service';
 
 const READ_ONLY_TYPES = ['form', 'translations'];
-const READ_ONLY_IDS = ['resources', 'branding', 'service-worker-meta', 'zscore-charts', 'settings', 'partners'];
+const READ_ONLY_IDS = ['resources', 'branding', DOC_IDS.SERVICE_WORKER_META, 'zscore-charts', 'settings', 'partners'];
```
Are these other `READ_ONLY_IDS` candidates for constants?

dianabarsan commented on sentinel/src/lib/metadata.js:
```diff
@@ -1,8 +1,9 @@
-const TRANSITION_SEQ_DOCUMENT = '_local/transitions-seq';
-const BACKGROUND_CLEANUP_SEQ_DOCUMENT = '_local/background-seq';
+const TRANSITION_SEQ_DOCUMENT = METADATA_KEYS.TRANSITIONS_SEQ;
+const BACKGROUND_CLEANUP_SEQ_DOCUMENT = METADATA_KEYS.BACKGROUND_SEQ;
```
These constants are now redundant

dianabarsan reviewed (commented):
Very nice! I think you're correct about the PR size, I appreciate you holding back from adding more changes. 
I left some really minor comments below.

cliftonmcintosh commented on api/src/migrations/extract-transition-seq.js:
```diff
@@ -4,9 +4,10 @@
-const TRANSITION_SEQ_DOCUMENT = '_local/transitions-seq';
-const BACKGROUND_CLEANUP_SEQ_DOCUMENT = '_local/background-seq';
+const TRANSITION_SEQ_DOCUMENT = METADATA_KEYS.TRANSITIONS_SEQ;
+const BACKGROUND_CLEANUP_SEQ_DOCUMENT = METADATA_KEYS.BACKGROUND_SEQ;
```
Updated

cliftonmcintosh commented:
> Very nice! I think you're correct about the PR size, I appreciate you holding back from adding more changes. I left some really minor comments below.

Thanks. Based on that, I will not add more constants in this PR. What is the best way to track adding more constants?

1. An issue with some sub-issues that breaks it down to manageable parts?
2. Several issues, each of which does a manageable part?
3. Something else?

I wouldn't want to spend the rest of my life updating constants, but I'm happy to plug away at the others bit by bit. I'm also happy to let someone else take them on or share more changes with me if you have someone who is interested.

mrjones-plip commented:
> I wouldn't want to spend the rest of my life updating constants, but I'm happy to plug away at the others bit by bit. I'm also happy to let someone else take them on or share more changes with me if you have someone who is interested.

Based on being a new comer - @cliftonmcintosh - do you think we could fashion a [good first issue](https://docs.communityhealthtoolkit.org/community/contributing/creating-good-first-issues/) out of another ticket?  I kind of like the idea of an iterative approach to this.  Each iteration can be inspired and informed by the last effort since they're very derivative of each other.  

However, since I'm not actually that close to the code, getting your (and @dianabarsan's!) take on this idea would be helpful.  Thanks!

cliftonmcintosh commented:
Compilation is failing because the constants can't be found in the admin app.

```sh
22 11 2025 00:36:35.039:INFO [Chrome Headless 142.0.0.0 (Linux x86_64)]: Connected on socket A4OmvVOkulvNwL7hAAAB with id 45696607
Chrome Headless 142.0.0.0 (Linux x86_64) ERROR
  Uncaught Error: Cannot find module '@medic/constants'
  at /home/runner/work/cht-core/cht-core/api/build/static/admin/js/main.js:1:206
  
  Error: Cannot find module '@medic/constants'
      at o (/home/runner/work/cht-core/cht-core/api/build/static/admin/js/main.js:1:166)
      at o (/home/runner/work/cht-core/cht-core/api/build/static/admin/js/main.js:1:132)
      at tests/unit/controllers/display-languages.spec.js:1:23
```

Compiling JS apps isn't my strongest skill.

Do I need to add this to the "dependencies" in `admin/package.json`? Is there a different thing I need to be doing?

```js
"@medic/constants": "file:../shared-libs/constants"
```

cliftonmcintosh commented:
> do you think we could fashion a [good first issue](https://docs.communityhealthtoolkit.org/community/contributing/creating-good-first-issues/) out of another ticket? I kind of like the idea of an iterative approach to this. Each iteration can be inspired and informed by the last effort since they're very derivative of each other.

Yes, as a first timer, I think having the continuation as first issues would work. One possibility would be to define categories of magic strings, like "document ids" or "document types" or "HTTP headers", and create a ticket for a category.

cliftonmcintosh commented:
# Admin module compilation error

> What I ran locally was I ran `cd admin && npm install ../shared-libs/constants`.

Thanks. I've tried this and then run the unit tests, but I still end up with an error that says

```
25 11 2025 13:30:36.194:WARN [filelist]: All files matched by "/Users/cliftonmcintosh/Documents/workspace/cht-core/node_modules/chai/chai.js" were excluded or matched by prior matchers.
25 11 2025 13:30:36.243:INFO [karma-server]: Karma v6.4.4 server started at http://localhost:9876/
25 11 2025 13:30:36.243:INFO [launcher]: Launching browsers Chrome_Headless with concurrency unlimited
25 11 2025 13:30:36.246:INFO [launcher]: Starting browser Chrome
25 11 2025 13:30:37.199:INFO [Chrome Headless 142.0.0.0 (Mac OS 10.15.7)]: Connected on socket 1u3gpGjOPGU4p_G4AAAB with id 52802022
Chrome Headless 142.0.0.0 (Mac OS 10.15.7) ERROR
  Uncaught Error: Cannot find module '@medic/constants'
  at /Users/cliftonmcintosh/Documents/workspace/cht-core/api/build/static/admin/js/main.js:1:206

  Error: Cannot find module '@medic/constants'
      at o (/Users/cliftonmcintosh/Documents/workspace/cht-core/api/build/static/admin/js/main.js:1:166)
      at o (/Users/cliftonmcintosh/Documents/workspace/cht-core/api/build/static/admin/js/main.js:1:132)
      at tests/unit/controllers/display-languages.spec.js:1:23
```

I did not commit the changes from running `npm install ../shared-libs/constants` because the unit tests still fail with that error.

<details>

<summary>Here is the diff after running `npm install ../shared-libs/constants`</summary>

```sh
diff --git a/admin/package-lock.json b/admin/package-lock.json
index 92f3dec68..6175b8e63 100644
--- a/admin/package-lock.json
+++ b/admin/package-lock.json
@@ -8,6 +8,7 @@
       "name": "medic-admin",
       "version": "0.1.0",
       "dependencies": {
+        "@medic/constants": "file:../shared-libs/constants",
         "@uirouter/angularjs": "^1.1.1",
         "angular": "^1.8.3",
         "angular-cookie": "^4.1.0",
@@ -45,6 +46,11 @@
       "extraneous": true,
       "license": "Apache-2.0"
     },
+    "../shared-libs/constants": {
+      "name": "@medic/constants",
+      "version": "1.0.0",
+      "license": "Apache-2.0"
+    },
     "../shared-libs/contact-types-utils": {
       "name": "@medic/contact-types-utils",
       "version": "1.0.0",
@@ -174,6 +180,10 @@
         "node": ">=6.9.0"
       }
     },
+    "node_modules/@medic/constants": {
+      "resolved": "../shared-libs/constants",
+      "link": true
+    },
     "node_modules/@types/linkify-it": {
       "version": "5.0.0",
       "resolved": "https://registry.npmjs.org/@types/linkify-it/-/linkify-it-5.0.0.tgz",
@@ -1059,6 +1069,9 @@
         "to-fast-properties": "^2.0.0"
       }
     },
+    "@medic/constants": {
+      "version": "file:../shared-libs/constants"
+    },
     "@types/linkify-it": {
       "version": "5.0.0",
       "resolved": "https://registry.npmjs.org/@types/linkify-it/-/linkify-it-5.0.0.tgz",
diff --git a/admin/package.json b/admin/package.json
index fee672748..c6b676de0 100644
--- a/admin/package.json
+++ b/admin/package.json
@@ -3,6 +3,7 @@
   "private": true,
   "version": "0.1.0",
   "dependencies": {
+    "@medic/constants": "file:../shared-libs/constants",
     "@uirouter/angularjs": "^1.1.1",
     "angular": "^1.8.3",
     "angular-cookie": "^4.1.0",
```

</details>

# String map error

> to fix your original error of `String map expecing string key`, it appears that ng-annotate has trouble with the `const { DOC_TYPES } = require(...)` Changing the require to:
> 
> ```
> const constants = require('@medic/constants');
> const DOC_TYPES = constants.DOC_TYPES;
> ```
> 
> fixes it.

I've adopted this approach for that String map error.

dianabarsan commented:
> Thanks. I've tried this and then run the unit tests, but I still end up with an error that says

The unit tests don't actually "build" the application. can you try running "npm run build-dev" before running the tests?

cliftonmcintosh commented:
> can you try running "npm run build-dev" before running the tests?

Thanks. That by itself didn't work. What does work for me is also adding `@medic/constants` to the browserify script so that it is bundled in and available when test files require it.

cliftonmcintosh commented on admin/tests/unit/controllers/display-languages.spec.js:
```diff
@@ -1,4 +1,5 @@
 describe('Display Languages controller', function() {
+  const { DOC_TYPES } = require('@medic/constants');
```
I moved this `require` statement inside of the describe block because when it was outside the describe block, I got a conflict between this declaration and the one in `admin/tests/unit/controllers/display-translations.spec.js`.

The clash looked like this:

```
  Uncaught SyntaxError: Identifier 'DOC_TYPES' has already been declared
  at tests/unit/controllers/display-translations.spec.js:1:1

  SyntaxError: Identifier 'DOC_TYPES' has already been declared
```

cliftonmcintosh replied:
Thanks for the explanation. 

I have a slight preference for the destructuring in the import statement because I feel it makes the access of the constants in the file just a little more concise.

However, if you feel that something like `CONSTANTS.DOC_TYPES.TRANSLATIONS` is easier to read or reason about than `DOC_TYPES.TRANSLATIONS`, I'm happy to change it.

cliftonmcintosh commented on admin/tests/unit/controllers/display-languages.spec.js:
```diff
@@ -58,15 +60,15 @@ describe('Display Languages controller', function() {
             _id: 'messages-en',
             code: 'en',
-            type: 'translations',
+            type: DOC_TYPES.TRANSLATIONS,
```
Oops. I missed the use of `'translations'` in the controller files. Thanks for noticing. I've updated them.

I also could not use destructuring for the imports for those files, so they are like this:

```js
const constants = require('@medic/constants');
const DOC_TYPES = constants.DOC_TYPES;
```

Please let me know if you think it would be better to adopt that same approach across the other files to be consistent.

Also please let me know if you think it's better to use `constants.DOC_TYPES.TRANSLATIONS` where the constant is being used instead of doing `const DOC_TYPES = constants.DOC_TYPES;` at the top of the file.

cliftonmcintosh commented:
@dianabarsan 

I think I've addressed your suggestions and comments. I've also resolved the conflicts with the main branch. Please let me know if you have advice for any other changes we should make in this.

dianabarsan commented:
There's nothing extra to do. There was a failing (flaking) e2e test that I had to restart to get the build green (otherwise merging is blocked). Otherwise I would have merged immediately after the approval :)
